File: sdk/ios/Sources/DleSDK/DleEvent.swift
import Foundation

/// Kind of a client event. Raw values are the wire names accepted by `POST /v1/events`
/// (`SdkEventNames`, spec §B.7.2).
public enum DleEventType: String, Sendable, Hashable, CaseIterable, Codable {
    /// The application was opened through one of the engine's links (spec §B.6.4, FR-223).
    /// This is the event that makes a direct open measurable at all.
    case linkOpen = "link_open"
    /// First launch after an installation.
    case firstOpen = "first_open"
    /// A foreground session started.
    case session = "session"
    /// A conversion, optionally carrying a monetary value.
    case conversion = "conversion"
    /// Anything the host application defines for itself.
    case custom = "custom"

    /// `true` for the types that describe user behaviour and are therefore gated on
    /// ``DleConsent/analytics``. `link_open` is not: it reports a URL the operating system
    /// handed to the app, which the engine treats as a first-party observation.
    public var isBehavioural: Bool {
        self != .linkOpen
    }
}

/// A single event for `POST /v1/events`. Mirrors `EventDto` member for member.
///
/// Keep `properties` free of personal data: the engine stores them as given.
public struct DleEvent: Sendable, Hashable, Codable {
    /// Event type.
    public var type: DleEventType
    /// Name of a conversion or custom event, for example `purchase`.
    public var name: String?
    /// URL that opened the application, for a `link_open` event.
    public var url: String?
    /// Monetary value of a conversion. Non-finite values are not sent.
    public var value: Double?
    /// ISO 4217 currency code accompanying ``value``, for example `EUR`.
    public var currency: String?
    /// When the event happened on the device. Filled in at enqueue time when `nil`.
    public var timestamp: Date?
    /// Flat string properties.
    public var properties: [String: String]?

    /// Creates an event.
    public init(
        type: DleEventType,
        name: String? = nil,
        url: String? = nil,
        value: Double? = nil,
        currency: String? = nil,
        timestamp: Date? = nil,
        properties: [String: String]? = nil
    ) {
        self.type = type
        self.name = name
        self.url = url
        self.value = value
        self.currency = currency
        self.timestamp = timestamp
        self.properties = properties
    }

    /// A `link_open` report for a URL the operating system handed to the app (FR-223).
    /// The fragment is dropped: it never reaches a server and can carry client-side state.
    public static func linkOpen(url: URL, at timestamp: Date? = nil) -> DleEvent {
        DleEvent(type: .linkOpen, url: DleUniversalLink.reportable(url), timestamp: timestamp)
    }

    /// A `session` event.
    public static func session(at timestamp: Date? = nil, properties: [String: String]? = nil) -> DleEvent {
        DleEvent(type: .session, timestamp: timestamp, properties: properties)
    }

    /// A `conversion` event.
    public static func conversion(
        name: String,
        value: Double? = nil,
        currency: String? = nil,
        at timestamp: Date? = nil,
        properties: [String: String]? = nil
    ) -> DleEvent {
        DleEvent(type: .conversion, name: name, value: value, currency: currency, timestamp: timestamp, properties: properties)
    }

    /// A `custom` event.
    public static func custom(name: String, at timestamp: Date? = nil, properties: [String: String]? = nil) -> DleEvent {
        DleEvent(type: .custom, name: name, timestamp: timestamp, properties: properties)
    }

    private enum CodingKeys: String, CodingKey {
        case type, name, url, value, currency, ts, properties
    }

    /// Encodes the `EventDto` shape: snake_case keys in contract order, absent members omitted
    /// rather than written as `null`, `ts` as an ISO 8601 string.
    public func encode(to encoder: Encoder) throws {
        var c = encoder.container(keyedBy: CodingKeys.self)
        try c.encode(type, forKey: .type)
        try c.encodeIfPresent(name, forKey: .name)
        try c.encodeIfPresent(url, forKey: .url)
        if let value, value.isFinite {
            try c.encode(value, forKey: .value)
        }
        try c.encodeIfPresent(currency, forKey: .currency)
        if let timestamp {
            try c.encode(DleWireDate.string(from: timestamp), forKey: .ts)
        }
        try c.encodeIfPresent(properties, forKey: .properties)
    }

    /// Decodes the `EventDto` shape. An unparsable `ts` becomes `nil` ("on arrival").
    public init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        type = try c.decode(DleEventType.self, forKey: .type)
        name = try c.decodeIfPresent(String.self, forKey: .name)
        url = try c.decodeIfPresent(String.self, forKey: .url)
        value = try c.decodeIfPresent(Double.self, forKey: .value)
        currency = try c.decodeIfPresent(String.self, forKey: .currency)
        if let raw = try c.decodeIfPresent(String.self, forKey: .ts) {
            timestamp = DleWireDate.date(from: raw)
        } else {
            timestamp = nil
        }
        properties = try c.decodeIfPresent([String: String].self, forKey: .properties)
    }
}

File: sdk/android/dle-sdk/src/main/kotlin/sk/magors/dle/DleEvent.kt
package sk.magors.dle

/**
 * Event kinds accepted by `POST /v1/events` (`SdkEventNames`, spec §B.7.2).
 *
 * @property wireName the exact string sent as `type`.
 * @property requiresAnalyticsConsent whether [DleClient.track] drops the event without
 * [DleConsent.analytics]. `link_open` is exempt: the engine records a direct open as first party
 * measurement of the link itself, no identifier crosses a context, and FR-223 requires the report.
 */
public enum class DleEventType(
    public val wireName: String,
    public val requiresAnalyticsConsent: Boolean,
) {
    /**
     * The application was opened through one of our links (spec §B.6.4, FR-223). The most
     * important event there is: a verified App Link opens the application without any request
     * reaching the engine, so this report is the only evidence the click ever happened.
     */
    LINK_OPEN("link_open", requiresAnalyticsConsent = false),

    /** First launch after an installation. Sent automatically once, after the first resolve. */
    FIRST_OPEN("first_open", requiresAnalyticsConsent = true),

    /** A foreground session started. */
    SESSION("session", requiresAnalyticsConsent = true),

    /** A conversion, optionally carrying a monetary value. */
    CONVERSION("conversion", requiresAnalyticsConsent = true),

    /** Anything the host application defines for itself. */
    CUSTOM("custom", requiresAnalyticsConsent = true),
    ;

    public companion object {
        /**
         * Looks a wire value up.
         *
         * @param value the `type` string.
         * @return the type, or `null` when unknown.
         */
        @JvmStatic
        public fun fromWireName(value: String): DleEventType? = entries.firstOrNull { it.wireName == value }
    }
}

/**
 * A single event queued for `POST /v1/events`. Mirrors `EventDto` (spec §B.7.2). Build instances
 * through the factories in the companion object; they enforce what the server would otherwise
 * reject.
 *
 * Keep [properties] free of personal data: the engine stores them as sent, and the deletion
 * endpoint keyed on `install_id` is the only remedy afterwards (spec §E.6.3).
 *
 * @property type the event kind.
 * @property name name of a conversion or custom event, for example `purchase`.
 * @property url URL that opened the application, for a [DleEventType.LINK_OPEN] event.
 * @property value monetary value of a conversion.
 * @property currency ISO 4217 currency code accompanying [value].
 * @property timestampMillis when the event happened on the device, in epoch milliseconds.
 * `null` is stamped at enqueue time.
 * @property properties additional flat string properties.
 */
public data class DleEvent(
    public val type: DleEventType,
    public val name: String? = null,
    public val url: String? = null,
    public val value: Double? = null,
    public val currency: String? = null,
    public val timestampMillis: Long? = null,
    public val properties: Map<String, String>? = null,
) {
    init {
        require(type != DleEventType.CONVERSION && type != DleEventType.CUSTOM || !name.isNullOrBlank()) {
            "a ${type.wireName} event needs a name"
        }
        require(type != DleEventType.LINK_OPEN || !url.isNullOrBlank()) { "a link_open event needs a url" }
        require(value == null || value.isFinite()) { "value must be a finite number" }
        require(currency == null || currency.length == CURRENCY_LENGTH) { "currency must be an ISO 4217 code" }
        require(properties == null || properties.size <= MAX_PROPERTIES) { "at most $MAX_PROPERTIES properties" }
    }

    override fun toString(): String = "DleEvent(type=${type.wireName}, name=$name)"

    public companion object {
        /** Length of an ISO 4217 currency code. */
        public const val CURRENCY_LENGTH: Int = 3

        /** Upper bound on [properties] entries, so a batch can never approach the server's body limit. */
        public const val MAX_PROPERTIES: Int = 32

        /**
         * The application was opened through a link. [DleClient.handleIntent] creates this for
         * you; call it directly only when you route intents yourself.
         *
         * @param url the URL the application was opened with. Send the scheme, host and path;
         * the query string is not needed and may carry personal data.
         * @param timestampMillis when it happened; `null` means now.
         */
        @JvmStatic
        @JvmOverloads
        public fun linkOpen(url: String, timestampMillis: Long? = null): DleEvent =
            DleEvent(type = DleEventType.LINK_OPEN, url = url, timestampMillis = timestampMillis)

        /** First launch after installation. */
        @JvmStatic
        public fun firstOpen(): DleEvent = DleEvent(type = DleEventType.FIRST_OPEN)

        /** A foreground session started. */
        @JvmStatic
        public fun session(): DleEvent = DleEvent(type = DleEventType.SESSION)

        /**
         * A conversion.
         *
         * @param name what converted, for example `purchase`.
         * @param value monetary value, when there is one.
         * @param currency ISO 4217 code such as `EUR`, upper cased for you.
         * @param properties additional flat string properties.
         */
        @JvmStatic
        @JvmOverloads
        public fun conversion(
            name: String,
            value: Double? = null,
            currency: String? = null,
            properties: Map<String, String>? = null,
        ): DleEvent = DleEvent(
            type = DleEventType.CONVERSION,
            name = name.trim(),
            value = value,
            currency = currency?.trim()?.uppercase(),
            properties = properties,
        )

        /**
         * An application defined event.
         *
         * @param name the event name.
         * @param properties additional flat string properties.
         */
        @JvmStatic
        @JvmOverloads
        public fun custom(name: String, properties: Map<String, String>? = null): DleEvent =
            DleEvent(type = DleEventType.CUSTOM, name = name.trim(), properties = properties)
    }
}
